release/calculate.go

```diff
@@ -36,70 +37,101 @@ func Calculate(finder ModuleFinder, tags git.ModuleTags, config repotools.Config
+			if err != nil {
+				return nil, fmt.Errorf("failed to list tombstone module files, %w", err)
+			}
+
+			files, err = gomod.FilterModuleFiles(module, files)
+			if err != nil {
+				return nil, fmt.Errorf("failed to filter tombstone module files, %w", err)
+			}
+
+			if len(files) != 0 {
+				return nil, fmt.Errorf("tombstone module has go source files, %v", files)
+			}
+			continue
+		}
+
+		moduleFile, err := gomod.LoadModuleFile(module.AbsPath(), nil, true)
 		if err != nil {
 			return nil, fmt.Errorf("failed to load module file: %w", err)
 		}
-
 		modulePath, err := gomod.GetModulePath(moduleFile)
 		if err != nil {
 			return nil, fmt.Errorf("failed to read module path: %w", err)
 		}
 
-		var latestVersion string
-		var hasChanges bool
-
-		latestVersion, ok := tags.Latest(moduleDir)
+		latestVersion, ok := tags.Latest(module.Path())
 		if ok {
-			startTag, err := git.ToModuleTag(moduleDir, latestVersion)
+			startTag, err := git.ToModuleTag(module.Path(), latestVersion)
 			if err != nil {
 				log.Fatalf("failed to convert module path and version to tag: %v", err)
 			}
 
-			changes, err := git.Changes(finder.Root(), startTag, "HEAD", moduleDir)
+			changes, err = git.Changes(finder.Root(), startTag, "HEAD", module.Path())
 			if err != nil {
 				log.Fatalf("failed to get git changes: %v", err)
 			}
 
-			subModulePaths := repositoryModules[moduleDir]
-
-			ignoredModulePaths := make([]string, 0, len(subModulePaths)+len(repositoryModuleTombstonePaths))
-			ignoredModulePaths = append(ignoredModulePaths, subModulePaths...)
-
-			if len(repositoryModuleTombstonePaths) > 0 {
-				ignoredModulePaths = append(ignoredModulePaths, repositoryModuleTombstonePaths...)
-				// IsModuleChanged expects the provided list of ignored modules paths to be sorted
-				sort.Strings(ignoredModulePaths)
-			}
-
-			hasChanges, err = gomod.IsModuleChanged(moduleDir, ignoredModulePaths, changes)
+			// Only consider changes that are specific to this module. Other
+			// module changes will be considered separately.
+			changes, err = gomod.FilterModuleFiles(module, changes)
 			if err != nil {
 				return nil, fmt.Errorf("failed to determine module changes: %w", err)
 			}
+			hasChanges = len(changes) != 0
 
 			if !hasChanges {
-				// Check if any of the submodules have been "carved out" of this module since the last tagged release
-				for _, subModuleDir := range subModulePaths {
-					if _, ok := tags.Latest(subModuleDir); ok {
+				// Check if any of the submodules have been "carved out" of
+				// this module since the last tagged release
+				for it := module.Iterator(); ; {
+					subModule := it.Next()
+					if subModule == nil {
+						break
+					}
+
+					// Ignore Tombstoned modules, since they no longer exist locally.
+					if module.HasAttribute(tombstonedModuleAttrib) {
+						continue
+					}
+
+					// Is a new submodule?
+					//  - yes, skip existing module
+					//  - no, check if carve out module
```

Comment:
nit: I believe the no/yes are flipped but the code is correct.

Should be:
```go
// Is a new submodule?
// - no, skip existing module
// - yes, check if carve out
```

Reply:
Thanks fixed wording here.